Track GPU memory usage

tsherif opened the pull request:
- Create a global `statsManager` singleton that allows independent objects to add their contributions to given stats.
- Track memory usage of buffers, textures and renderbuffers
- Add a memory stats widget to the 3D texture example

Key design question I wouldn't mind getting some feedback on is the `StatsManager`. Couldn't think of another way to keep memory usage stats in one place. Let me know if you guys can think of alternatives.

ibgreen commented on modules/core/src/webgl/classes/renderbuffer.js:
```diff
@@ -29,13 +30,22 @@ export default class Renderbuffer extends Resource {
 
   constructor(gl, opts = {}) {
     super(gl, opts);
+    this.gpuMemoryStats = statsManager.get('Memory Usage').get('GPU Memory');
```
Would it make sense to define constants for these strings and decouple them from the printable representation? Or would that create too much coupling in the code?

Pessimistress replied:
Or this could be done in the base `Resource` class constructor. We may also add a method `resource.setMemoryUsage` or something.

tsherif replied:
Yeah, it would make sense to move at least parts of this into `Resource`. With the `byteLength` naming convention, it could easily manage everything except the per-type stats.

ibgreen commented on modules/core/src/webgl/classes/renderbuffer.js:
```diff
@@ -90,4 +108,38 @@ export default class Renderbuffer extends Resource {
   }
+
+  /* eslint-disable complexity */
+  _getFormatSize() {
```
Idea: We already have a table for formats, maybe reuse that or add this field to that table?

See top of file:
```
import RENDERBUFFER_FORMATS from './renderbuffer-formats';
```

tsherif replied:
Good call. And since the formats are the same between textures and renderbuffers, I could have one for both.

ibgreen commented on modules/core/src/webgl/classes/texture-3d.js:
```diff
@@ -15,4 +18,76 @@ export default class Texture3D extends Texture {
+    data,
+    parameters = {}
+  }) {
+    this.gpuMemoryStats.subtractCount(this.byteLength);
```
Maybe call super() and then add stats instead of moving here?

ibgreen commented on modules/core/src/webgl/classes/texture-formats.js:
```diff
@@ -134,3 +134,48 @@ export function isLinearFilteringSupported(gl, format) {
+  // TODO(Tarek): Cover all formats/types?
+  const numTexels = width * height * (depth || 1);
+  let channels;
+  switch (dataFormat) {
```
Again worth looking if we can reuse existing tables?

ibgreen commented on modules/core/src/webgl/classes/buffer.js:
```diff
@@ -3,6 +3,7 @@ import Resource from './resource';
 import Accessor from './accessor';
 import {assertWebGL2Context, getGLTypeFromTypedArray, getTypedArrayFromGLType} from '../utils';
 import {log, assert, checkProps} from '../../utils';
+import statsManager from '../../core/stats-manager';
```
Note: Buffers can be `reallocated`, also check `setData`/`_setData` to follow how they reinitialize buffers. Might want to keep things updated.

tsherif replied:
So I believe this is covered by the fact it always subtracts the current `byteLength` from the counts before adding to the count after allocation. Is there a path that I'm missing?

ibgreen replied:
OK. Worth double checking `Buffer._setByteLength` I think it can change the size of the buffer.

tsherif replied:
Ah good catch! Wasn't covering the `reallocate` method. Moving tracking into `_setData` and `_setByteLength`

tsherif commented:
@ibgreen can you take another look at this? Primary changes are:
- Move total GPU memory tracking into `Resource` methods `_trackDeallocatedMemory` and `_trackAllocatedMemory`, subclasses override to track type-specific stats.
- Added `pixelSize`, `texelSize`, `channels` entries to `TEXTURE_FORMATS` and `RENDERBUFFER_FORMATS` tables to be used for size calculations.
- Filled out entries in the `TEXTURE_FORMATS` table to ensure coverage of most types we might use.
- `TYPE_SIZES` table in `texture-formats` module, used for size calculations.
- Fixed specs to handle new formats in the `TEXTURE_FORMATS` table.

ibgreen commented on examples/core/3d-texture/app.js:
```diff
@@ -152,12 +152,27 @@ class AppAnimationLoop extends AnimationLoop {
     });
 
-    return {cloud, mvpMat, viewMat, statsWidget};
+    const memWidget = new StatsWidget(statsManager.get('Memory Usage'), {
```
Side note:

My proposal is that we cut this widget from the individual examples and add it to the website.

That way we get the benefits in all demos without cluttering up every app.

I made a big push on the website in the pending scenegraph split PR, should be able to go through how it would work once that lands.

ibgreen commented on examples/core/instancing/app.js:
```diff
@@ -154,13 +162,36 @@ class AppAnimationLoop extends AnimationLoop {
     });
 
-    return {statsWidget};
+    const memWidget = new StatsWidget(statsManager.get('Memory Usage'), {
```
Side note: My initial thinking is that we would have one widget that can show multiple screens and can be expanded/collapsed/circulated by the user, rather than add multiple widgets. Zero config in apps and all power in the users hands to slice and dice the data. Such sophistication would come later of course.

ibgreen commented on modules/core/src/webgl/classes/renderbuffer-formats.js:
```diff
@@ -7,56 +7,56 @@ const EXT_FLOAT_WEBGL2 = 'EXT_color_buffer_float';
-  [GL.RG32I]: {gl2: true},
-  [GL.RGB8]: {gl2: true},
-  [GL.RGBA8]: {gl2: true},
+  [GL.R8]: {gl2: true, bytesPerPixel: 1},
```
Tables take a fair bit of space in the bundle, I have always had some angst about these. And I believe that key strings i.e. `{bytesPerPixel: ...}` are usually not minified like variable names (worth checking). I'm normally a big proponent of typing things out but in this case maybe something shorter (`bpp`?), or use arrays and destructure on use.

tsherif replied:
I assumed the minification would work if it's not exported externally? Not something I know too well though, so I'll just change it to `bpp`

ibgreen commented on modules/core/src/webgl/classes/renderbuffer.js:
```diff
@@ -29,13 +30,20 @@ export default class Renderbuffer extends Resource {
 
   constructor(gl, opts = {}) {
     super(gl, opts);
+    this.renderbufferMemoryStats = statsManager.get('Memory Usage').get('Renderbuffer Memory');
```
Is this lookup an cheap operation? If so, maybe not save this member on the object?

ibgreen commented on modules/core/src/webgl/classes/resource.js:
```diff
@@ -284,4 +289,14 @@ export default class Resource {
     stats.resourceMap[name].active--;
   }
+
+  _trackAllocatedMemory(bytes) {
```
Nit: Probably a lot of extra work, but assume it could be valuable to see what type of resource is using the bytes? I.e. add one more param?

tsherif replied:
That's what's happening in the overrides in the subclasses.

ibgreen commented on modules/core/src/webgl/classes/texture-formats.js:
```diff
@@ -11,50 +11,153 @@ import {isWebGL2} from '../utils';
-  // Sized texture format - more performance
+  // Sized/compressed
   // R
-  // [GL.R8]: {dataFormat: GL.RED, types: [GL.UNSIGNED_BYTE], gl2: true},
```
So this was commented out for bundle size reasons. Would be good if you could run collect metrics before and after this change. And make a pass on making the table smaller. That could be a second PR, no need to hold this one, but let's not just add this without checking.

tsherif replied:
Ok, sure. I'm thinking what I'll do is create another table `DATA_FORMAT_CHANNELS`, which will be quite small, and then I can do the calculations for most cases based on `numChannels * typeSize`.

Might not be an issue for our current use cases, but it seems odd that we're returning "not supported" for a lot of useful WebGL 2 formats

ibgreen commented on modules/core/src/webgl/classes/texture-formats.js:
```diff
@@ -108,16 +211,31 @@ export const TEXTURE_FORMATS = {
     return true;
   }
-  const value = isWebGL2(gl) ? info.gl2 || info.gl1 : info.gl1;
+  const value = isWebGL2(gl) ? info.gl2 : info.gl1;
```
I believe this was written this way to save space and not duplicate gl2 entries if gl1 was identical. Clearly there should have been comments indicating this.

tsherif replied:
The issue I had was formats that exist in WebGL 1 but not 2, mainly the depth texture extension formats.

ibgreen commented on modules/core/src/webgl/classes/texture.js:
```diff
@@ -315,6 +320,18 @@ export default class Texture extends Resource {
+    if (data && data.byteLength) {
+      this._trackAllocatedMemory(data.byteLength);
+    } else {
+      let bytesPerTexel = TEXTURE_FORMATS[this.format].bytesPerTexel;
```
Maybe helper methods to calculate this from the table, both here and in other file?

Pessimistress commented on modules/core/src/core/stats-manager.js:
```diff
@@ -0,0 +1,17 @@
+import {Stats} from 'probe.gl';
+
+class StatsManager {
```
There is a global object `luma.stats` that tracks the count of resource objects by type. Worth consolidating?

tsherif replied:
Definitely! Hadn't notice that.

tsherif commented:
@ibgreen @Pessimistress Take 3!
- Consolidate stat tracking in global `luma.stats`, which is now a "stats manager"-like object.
- Revert `TEXTURE_FORMAT` table updates. Create a `DATA_FORMAT_CHANNELS` table that can be used with `TYPE_SIZES` to calculate byte-per-pixel for textures.
- Remove subclass `_trackAllocatedMemory` and `_trackDeallocatedMemory` methods using a pattern similar to what happens in `Resource._addStats`.

ibgreen commented on modules/core/src/core/animation-loop.js:
```diff
@@ -10,7 +10,7 @@ import {getPageLoadPromise} from '../webgl/context';
 import {log} from '../utils';
 import assert from '../utils/assert';
-import {Stats} from 'probe.gl';
+import luma from '../init';
```
If we import it like this, whenever we use it, we don't need to add it to `.eslintrc`?

tsherif replied:
The issue I had was in the examples, since the `luma` object isn't exported from Luma.gl. Should we export it?

ibgreen replied:
> Should we export it?

In that case we should probably export the `stats` object, not the luma wrapper object.

tsherif replied:
Sounds good.

ibgreen commented on modules/core/test/webgl/classes/texture.spec.js:
```diff
@@ -123,12 +123,18 @@ function testFormatCreation(t, glContext, withData = false) {
         format,
         type,
-        mipmaps: format !== GL.RGB32F, // TODO: for some reason mipmap generation failing for RGB32F format
+        mipmaps: formatInfo.dataFormat === GL.RGBA, // Avoid mipmapping compressed formats
```
Is `GL.RGBA` true for `GL.R` etc? Maybe a bit more comment here?

tsherif replied:
Ah, this is left over from when the TEXTURE_FORMATS table was expanded. I'll revert.
